improve URL supporting

Amazing feature, thank you! This allows us to use Wasm in npm packages that can be used in Web, Node and Webpack projects.

Did you consider not requiring `./` at the beginning of the relative path? Both

```js
let a = new URL("./wasm/drand.wasm", "file:///Users/me/projects/drand-client/verifying.js");
let b = new URL("wasm/drand.wasm", "file:///Users/me/projects/drand-client/verifying.js");
```

are the same URL but the second one fails with

> Module not found: Error: Can't resolve 'wasm/drand.wasm' in '/Users/simon/opensource/drand/drand-client'

_Originally posted by @webmaster128 in https://github.com/webpack/webpack/pull/11276#issuecomment-694070263_

---

This issue was moved from webpack/webpack#11491 by @vankop. Original issue was by @evilebottnawi.

--- a/README.md
+++ b/README.md
@@ -95,6 +95,7 @@ myResolver.resolve({}, lookupStartPath, request, resolveContext, (
 | plugins          | []                          | A list of additional resolve plugins which should be applied                                                                                              |
 | resolver         | undefined                   | A prepared Resolver to which the plugins are attached                                                                                                     |
 | resolveToContext | false                       | Resolve to a context instead of a file                                                                                                                    |
+| preferRelative   | false                       | Prefer to resolve module requests as relative request and fallback to resolving as module                                                                 |
 | restrictions     | []                          | A list of resolve restrictions                                                                                                                            |
 | roots            | []                          | A list of root paths                                                                                                                                      |
 | symlinks         | true                        | Whether to resolve symlinks to their symlinked location                                                                                                   |

--- a/lib/ResolverFactory.js
+++ b/lib/ResolverFactory.js
@@ -73,6 +73,7 @@ const UseFilePlugin = require("./UseFilePlugin");
  * @property {boolean=} resolveToContext Resolve to a context instead of a file
  * @property {(string|RegExp)[]=} restrictions A list of resolve restrictions
  * @property {boolean=} useSyncFileSystemCalls Use only the sync constiants of the file system calls
+ * @property {boolean=} preferRelative Prefer to resolve module requests as relative requests before falling back to modules
  */
 
 /**
@@ -101,6 +102,7 @@ const UseFilePlugin = require("./UseFilePlugin");
  * @property {boolean} fullySpecified
  * @property {boolean} resolveToContext
  * @property {Set<string|RegExp>} restrictions
+ * @property {boolean} preferRelative
  */
 
 /**
@@ -218,6 +220,7 @@ function createOptions(options) {
 		roots: new Set(options.roots || undefined),
 		fullySpecified: options.fullySpecified || false,
 		resolveToContext: options.resolveToContext || false,
+		preferRelative: options.preferRelative || false,
 		restrictions: new Set(options.restrictions)
 	};
 }
@@ -249,6 +252,7 @@ exports.createResolver = function (options) {
 		plugins: userPlugins,
 		pnpApi,
 		resolveToContext,
+		preferRelative,
 		symlinks,
 		unsafeCache,
 		resolver: customResolver,
@@ -338,6 +342,9 @@ exports.createResolver = function (options) {
 			new AliasFieldPlugin("normal-resolve", item, "internal-resolve")
 		);
 	});
+	if (preferRelative) {
+		plugins.push(new JoinRequestPlugin("after-normal-resolve", "relative"));
+	}
 	plugins.push(
 		new ConditionalPlugin(
 			"after-normal-resolve",
@@ -359,7 +366,9 @@ exports.createResolver = function (options) {
 	if (roots.size > 0) {
 		plugins.push(new RootsPlugin("after-normal-resolve", roots, "relative"));
 	}
-	plugins.push(new JoinRequestPlugin("after-normal-resolve", "relative"));
+	if (!preferRelative) {
+		plugins.push(new JoinRequestPlugin("after-normal-resolve", "relative"));
+	}
 
 	// internal
 	importsFields.forEach(importsField => {

--- a/test/resolve.js
+++ b/test/resolve.js
@@ -20,6 +20,10 @@ const issue238Resolve = resolve.create({
 	modules: ["src/a", "src/b", "src/common", "node_modules"]
 });
 
+const preferRelativeResolve = resolve.create({
+	preferRelative: true
+});
+
 function testResolve(name, context, moduleName, result) {
 	describe(name, function () {
 		it("should resolve sync correctly", function () {
@@ -250,5 +254,23 @@ describe("resolve", function () {
 				done();
 			}
 		);
+	});
+
+	it("should correctly resolve with preferRelative", function (done) {
+		preferRelativeResolve(fixtures, "main1.js", function (err, filename) {
+			if (err) done(err);
+			should.exist(filename);
+			filename.should.equal(path.join(fixtures, "main1.js"));
+			done();
+		});
+	});
+
+	it("should correctly resolve with preferRelative", function (done) {
+		preferRelativeResolve(fixtures, "m1/a.js", function (err, filename) {
+			if (err) done(err);
+			should.exist(filename);
+			filename.should.equal(path.join(fixtures, "node_modules", "m1", "a.js"));
+			done();
+		});
 	});
 });

--- a/test/resolve.js
+++ b/test/resolve.js
@@ -25,14 +25,14 @@ const preferRelativeResolve = resolve.create({
 });
 
 function testResolve(name, context, moduleName, result) {
-	describe(name, function () {
-		it("should resolve sync correctly", function () {
+	describe(name, function() {
+		it("should resolve sync correctly", function() {
 			var filename = resolve.sync(context, moduleName);
 			should.exist(filename);
 			filename.should.equal(result);
 		});
-		it("should resolve async correctly", function (done) {
-			resolve(context, moduleName, function (err, filename) {
+		it("should resolve async correctly", function(done) {
+			resolve(context, moduleName, function(err, filename) {
 				if (err) return done(err);
 				should.exist(filename);
 				filename.should.equal(result);
@@ -43,23 +43,23 @@ function testResolve(name, context, moduleName, result) {
 }
 
 function testResolveContext(name, context, moduleName, result) {
-	describe(name, function () {
-		it("should resolve async correctly", function (done) {
-			asyncContextResolve(context, moduleName, function (err, filename) {
+	describe(name, function() {
+		it("should resolve async correctly", function(done) {
+			asyncContextResolve(context, moduleName, function(err, filename) {
 				if (err) done(err);
 				should.exist(filename);
 				filename.should.equal(result);
 				done();
 			});
 		});
-		it("should resolve sync correctly", function () {
+		it("should resolve sync correctly", function() {
 			var filename = syncContextResolve(context, moduleName);
 			should.exist(filename);
 			filename.should.equal(result);
 		});
 	});
 }
-describe("resolve", function () {
+describe("resolve", function() {
 	testResolve(
 		"absolute path",
 		fixtures,
@@ -239,13 +239,13 @@ describe("resolve", function () {
 		path.join(fixtures, "main-field-self2", "index.js")
 	);
 
-	it("should correctly resolve", function (done) {
+	it("should correctly resolve", function(done) {
 		const issue238 = path.resolve(fixtures, "issue-238");
 
 		issue238Resolve(
 			path.resolve(issue238, "./src/common"),
 			"config/myObjectFile",
-			function (err, filename) {
+			function(err, filename) {
 				if (err) done(err);
 				should.exist(filename);
 				filename.should.equal(
@@ -256,8 +256,8 @@ describe("resolve", function () {
 		);
 	});
 
-	it("should correctly resolve with preferRelative", function (done) {
-		preferRelativeResolve(fixtures, "main1.js", function (err, filename) {
+	it("should correctly resolve with preferRelative", function(done) {
+		preferRelativeResolve(fixtures, "main1.js", function(err, filename) {
 			if (err) done(err);
 			should.exist(filename);
 			filename.should.equal(path.join(fixtures, "main1.js"));
@@ -265,8 +265,8 @@ describe("resolve", function () {
 		});
 	});
 
-	it("should correctly resolve with preferRelative", function (done) {
-		preferRelativeResolve(fixtures, "m1/a.js", function (err, filename) {
+	it("should correctly resolve with preferRelative", function(done) {
+		preferRelativeResolve(fixtures, "m1/a.js", function(err, filename) {
 			if (err) done(err);
 			should.exist(filename);
 			filename.should.equal(path.join(fixtures, "node_modules", "m1", "a.js"));

--- a/test/resolve.js
+++ b/test/resolve.js
@@ -25,14 +25,14 @@ const preferRelativeResolve = resolve.create({
 });
 
 function testResolve(name, context, moduleName, result) {
-	describe(name, function() {
-		it("should resolve sync correctly", function() {
+	describe(name, function () {
+		it("should resolve sync correctly", function () {
 			var filename = resolve.sync(context, moduleName);
 			should.exist(filename);
 			filename.should.equal(result);
 		});
-		it("should resolve async correctly", function(done) {
-			resolve(context, moduleName, function(err, filename) {
+		it("should resolve async correctly", function (done) {
+			resolve(context, moduleName, function (err, filename) {
 				if (err) return done(err);
 				should.exist(filename);
 				filename.should.equal(result);
@@ -43,23 +43,23 @@ function testResolve(name, context, moduleName, result) {
 }
 
 function testResolveContext(name, context, moduleName, result) {
-	describe(name, function() {
-		it("should resolve async correctly", function(done) {
-			asyncContextResolve(context, moduleName, function(err, filename) {
+	describe(name, function () {
+		it("should resolve async correctly", function (done) {
+			asyncContextResolve(context, moduleName, function (err, filename) {
 				if (err) done(err);
 				should.exist(filename);
 				filename.should.equal(result);
 				done();
 			});
 		});
-		it("should resolve sync correctly", function() {
+		it("should resolve sync correctly", function () {
 			var filename = syncContextResolve(context, moduleName);
 			should.exist(filename);
 			filename.should.equal(result);
 		});
 	});
 }
-describe("resolve", function() {
+describe("resolve", function () {
 	testResolve(
 		"absolute path",
 		fixtures,
@@ -239,13 +239,13 @@ describe("resolve", function() {
 		path.join(fixtures, "main-field-self2", "index.js")
 	);
 
-	it("should correctly resolve", function(done) {
+	it("should correctly resolve", function (done) {
 		const issue238 = path.resolve(fixtures, "issue-238");
 
 		issue238Resolve(
 			path.resolve(issue238, "./src/common"),
 			"config/myObjectFile",
-			function(err, filename) {
+			function (err, filename) {
 				if (err) done(err);
 				should.exist(filename);
 				filename.should.equal(
@@ -256,8 +256,8 @@ describe("resolve", function() {
 		);
 	});
 
-	it("should correctly resolve with preferRelative", function(done) {
-		preferRelativeResolve(fixtures, "main1.js", function(err, filename) {
+	it("should correctly resolve with preferRelative", function (done) {
+		preferRelativeResolve(fixtures, "main1.js", function (err, filename) {
 			if (err) done(err);
 			should.exist(filename);
 			filename.should.equal(path.join(fixtures, "main1.js"));
@@ -265,8 +265,8 @@ describe("resolve", function() {
 		});
 	});
 
-	it("should correctly resolve with preferRelative", function(done) {
-		preferRelativeResolve(fixtures, "m1/a.js", function(err, filename) {
+	it("should correctly resolve with preferRelative", function (done) {
+		preferRelativeResolve(fixtures, "m1/a.js", function (err, filename) {
 			if (err) done(err);
 			should.exist(filename);
 			filename.should.equal(path.join(fixtures, "node_modules", "m1", "a.js"));

--- a/types.d.ts
+++ b/types.d.ts
@@ -194,6 +194,7 @@ declare interface ResolveOptions {
 	fullySpecified: boolean;
 	resolveToContext: boolean;
 	restrictions: Set<string | RegExp>;
+	preferRelative: boolean;
 }
 declare abstract class Resolver {
 	fileSystem: FileSystem;
@@ -415,6 +416,11 @@ declare interface UserResolveOptions {
 	 * Use only the sync constiants of the file system calls
 	 */
 	useSyncFileSystemCalls?: boolean;
+
+	/**
+	 * Prefer to resolve module requests as relative requests before falling back to modules
+	 */
+	preferRelative?: boolean;
 }
 declare function exports(
 	context?: any,

